File: ops-agent/src/lib/chunker.py
import re


def chunk_text(text: str, max_chars: int = 500) -> list[str]:
    if not text or not text.strip():
        return []

    # Split by markdown headers or double newlines
    segments = re.split(r"\n(?=#{1,6}\s)|\n\n+", text)

    chunks: list[str] = []
    for segment in segments:
        segment = segment.strip()
        if not segment:
            continue

        if len(segment) <= max_chars:
            chunks.append(segment)
        else:
            # Split long segments by lines
            _split_long_segment(segment, max_chars, chunks)

    return chunks


def _split_long_segment(segment: str, max_chars: int, chunks: list[str]) -> None:
    lines = segment.split("\n")
    current = ""
    for line in lines:
        candidate = f"{current}\n{line}".strip() if current else line
        if len(candidate) <= max_chars:
            current = candidate
        else:
            if current:
                chunks.append(current)
            current = line[:max_chars] if len(line) > max_chars else line
    if current.strip():
        chunks.append(current)

File: server/src/lib/chunker.ts
export interface ParsedSegment {
  content: string;
  metadata: Record<string, unknown>;
}

export interface Chunk {
  content: string;
  metadata: Record<string, unknown>;
  index: number;
}

/**
 * 对每个 segment 分块并传播其 metadata，最终统一分配递增 index。
 */
export function chunkSegments(
  segments: ParsedSegment[],
  maxChars = 500,
): Chunk[] {
  const result: Chunk[] = [];
  let index = 0;
  for (const seg of segments) {
    const texts = chunkText(seg.content, maxChars);
    for (const text of texts) {
      result.push({ content: text, metadata: seg.metadata, index: index++ });
    }
  }
  return result;
}

/**
 * 将文本按 markdown 标题 / 双换行 / 行 分块。
 */
export function chunkText(text: string, maxChars = 500): string[] {
  if (!text?.trim()) return [];

  // 按 markdown header 或双换行分割
  const segments = text.split(/\n(?=#{1,6}\s)|\n\n+/);
  const chunks: string[] = [];

  for (const raw of segments) {
    const segment = raw.trim();
    if (!segment) continue;

    if (segment.length <= maxChars) {
      chunks.push(segment);
    } else {
      splitLongSegment(segment, maxChars, chunks);
    }
  }

  return chunks;
}

function splitLongSegment(
  segment: string,
  maxChars: number,
  chunks: string[],
): void {
  const lines = segment.split("\n");
  let current = "";

  for (const line of lines) {
    const candidate = current ? `${current}\n${line}`.trim() : line;
    if (candidate.length <= maxChars) {
      current = candidate;
    } else {
      if (current) chunks.push(current);
      current = line.length > maxChars ? line.slice(0, maxChars) : line;
    }
  }

  if (current.trim()) {
    chunks.push(current);
  }
}
